### PR TITLE
Ignoring some variables during autoscaling

### LINKED ISSUE #19572: Specify which variables are to be autoscaled
## Reason
Autoscaling is awesome...most of the time. Sometimes, a variable's residual is way off during a given timestep, such that manual scaling is more appropriate. Being able to choose which variables are available for automatic scaling, and others are only manually scaled, will allow flexibility.

## Design
An executioner option like `autoscaled_variables = 'temp disp_x disp_y` or something similar.

## Impact
Flexibility in autoscaling.

### PR DESCRIPTION
**Pull request by tophmatthews**
Closes #19572

### CODE REVIEW

**Comment by tophmatthews** [comment]
Here's a hack at this @lindsayad . One thing that I'm not sure of is if the scaling for the "non-ignored" variables should change if some variables are ignored. Currently, the non-ignored variables have the same scaling with and without all variables included. If feels like that shouldn't be the case...but I'm not positive.

File: framework/src/systems/NonlinearSystemBase.C [hunk 1]
```diff
@@ -3444,55 +3468,57 @@ NonlinearSystemBase::computeScaling()
   // Compute our scaling factors for the spatial field variables
   for (const auto & elem : *mesh().getActiveLocalElementRange())
     for (MooseIndex(field_variables) i = 0; i < field_variables.size(); ++i)
+      if (field_variable_autoscaled[i])
```
Comment (tophmatthews) [inline framework/src/systems/NonlinearSystemBase.C:3471]: Kinda hard to see...but the only change here is adding this new "if" statement. The rest is just formatting.

File: framework/src/executioners/FEProblemSolve.C [hunk 2]
```diff
@@ -113,6 +113,9 @@ FEProblemSolve::validParams()
       "scaling_group_variables",
       "Name of variables that are grouped together for determining scale factors. (Multiple "
       "groups can be provided, separated by semicolon)");
+  params.addParam<std::vector<std::string>>(
+      "ignore_variables_for_autoscaling",
+      "List of variables that do not participate in autoscaling.");
```
Comment (GiudGiud) [inline framework/src/executioners/FEProblemSolve.C:118]: needs to be added to the parameter group

File: framework/src/systems/NonlinearSystemBase.C [hunk 3]
```diff
@@ -3355,6 +3355,13 @@ NonlinearSystemBase::setupScalingGrouping()
          ++group_index)
       for (const auto & var_name : _scaling_group_variables[group_index])
       {
+        if (find(_ignore_variables_for_autoscaling.begin(),
+                 _ignore_variables_for_autoscaling.end(),
+                 var_name) != _ignore_variables_for_autoscaling.end())
+          mooseError("Variable ",
+                     var_name,
+                     " cannot be in a scaling grouping and be also ignored via "
+                     "'ignore_variables_for_autoscaling'");
```
Comment (lindsayad) [inline framework/src/systems/NonlinearSystemBase.C:3364]: I think you could do this check in `FEProblemSolve` and be able to give a `paramError` as opposed to a `mooseError`

File: framework/src/systems/NonlinearSystemBase.C [hunk 4]
```diff
@@ -3396,7 +3403,24 @@ NonlinearSystemBase::computeScaling()
...
+      if (find(_ignore_variables_for_autoscaling.begin(),
+               _ignore_variables_for_autoscaling.end(),
+               (*field_variables[i]).name()) != _ignore_variables_for_autoscaling.end())
+        field_variable_autoscaled[i] = false;
+    for (MooseIndex(scalar_variable_autoscaled) i = 0; i < scalar_variable_autoscaled.size(); ++i)
+      if (find(_ignore_variables_for_autoscaling.begin(),
+               _ignore_variables_for_autoscaling.end(),
+               (*scalar_variables[i]).name()) != _ignore_variables_for_autoscaling.end())
+        scalar_variable_autoscaled[i] = false;
+  }
```
Comment (lindsayad) [inline framework/src/systems/NonlinearSystemBase.C:3423]: I would rename `setupScalingGrouping` to something like `setupScalingData` and construct `_field_variable_autoscaled` and `_scalar_variable_autoscaled` data members in there. Then you only do these O(N) algorithms once. Also make sure to use `std::find` as opposed to just `find`, unless you are using some other `find` and then please explicitly apply that scoping operator instead

**Review by lindsayad — CHANGES_REQUESTED** [review]
Looking good

File: framework/src/executioners/FEProblemSolve.C [hunk 5]
```diff
@@ -218,6 +221,31 @@ FEProblemSolve::FEProblemSolve(Executioner & ex)
...
         getParam<std::vector<std::vector<std::string>>>("scaling_group_variables"));
+  if (isParamValid("ignore_variables_for_autoscaling"))
+  {
+    // Before setting ignore_variables_for_autoscaling, check that they are not present in
+    // scaling_group_variables
+    if (isParamValid("scaling_group_variables"))
+    {
+      const auto ignore_variables_for_autoscaling =
+          getParam<std::vector<std::string>>("ignore_variables_for_autoscaling");
+      const auto scaling_group_variables =
```
Comment (lindsayad) [inline framework/src/executioners/FEProblemSolve.C:232]:
```suggestion
      const auto & ignore_variables_for_autoscaling =
          getParam<std::vector<std::string>>("ignore_variables_for_autoscaling");
      const auto & scaling_group_variables =
```

File: framework/src/executioners/FEProblemSolve.C [hunk 6]
```diff
@@ -218,6 +221,31 @@ FEProblemSolve::FEProblemSolve(Executioner & ex)
...
+          getParam<std::vector<std::vector<std::string>>>("scaling_group_variables");
+      for (MooseIndex(scaling_group_variables) group_index = 0;
+           group_index < scaling_group_variables.size();
+           ++group_index)
+        for (const auto & var_name : scaling_group_variables[group_index])
```
Comment (lindsayad) [inline framework/src/executioners/FEProblemSolve.C:237]:
```suggestion
      for (const auto & group : scaling_group_variables)
        for (const auto & var_name : group)
```

File: framework/src/executioners/FEProblemSolve.C [hunk 7]
```diff
@@ -218,6 +221,31 @@ FEProblemSolve::FEProblemSolve(Executioner & ex)
...
+        {
+          if (std::find(ignore_variables_for_autoscaling.begin(),
+                        ignore_variables_for_autoscaling.end(),
+                        var_name) != ignore_variables_for_autoscaling.end())
+            paramError(" ignore_variables_for_autoscaling",
```
Comment (lindsayad) [inline framework/src/executioners/FEProblemSolve.C:242]:
```suggestion
            paramError("ignore_variables_for_autoscaling",
```

**Review by lindsayad — CHANGES_REQUESTED** [review]
As long as coverage looks good, I'm ready to merge after this

**Review by lindsayad — APPROVED** [review]

**Pull request merged**